org.eclipse.xtext/src/org/eclipse/xtext/EcoreUtil2.java

```diff
@@ -773,4 +775,58 @@ protected EObject computeNext() {
 			}
 		};
 	}
+	
+	public static URI getPathURI(final EObject object) {
+		Resource resource = object.eResource();
+		if(resource != null) {
+			String fragment = getPathURIFragment(object);
+			URI resourceURI = getPlatformResourceOrNormalizedURI(resource);
+			return resourceURI.appendFragment(fragment);
+		} else {
+			return getPlatformResourceOrNormalizedURI(object);
+		}
+	}
+
+	public static String getPathURIFragment(EObject object) {
+		SegmentSequence.Builder builder = SegmentSequence.newBuilder("/");
+		InternalEObject internalEObject = (InternalEObject) object;
+		boolean isContained = internalEObject.eDirectResource() != null;
+		for (InternalEObject container = internalEObject.eInternalContainer(); 
+			container != null && !isContained; 
+			container = internalEObject.eInternalContainer()) {
+			builder.append(getPathURIFragment(container, internalEObject.eContainingFeature(), internalEObject));
+			internalEObject = container;
+			if (container.eDirectResource() != null) {
+				isContained = true;
+			}
+		}
+		if (!isContained) {
+			return "/-1";
+		}
+		builder.append(getURIFragmentRootSegment(internalEObject));
+		builder.append("");
+		builder.reverse();
+		return builder.toSegmentSequence().toString();
+	}
+	
+	protected static String getPathURIFragment(InternalEObject container, EStructuralFeature feature, InternalEObject contained) {
```

Comment:
should this be named `getPathURISegment`?

Reply:
I have often heard the name _fragment path_ for fragments like this, so I changed it to this.